common/src/lib.rs

```diff
@@ -2,15 +2,79 @@ pub mod crypto;
 
 use crypto::*;
 use serde::{Deserialize, Serialize};
-use std::collections::BTreeSet;
+use std::collections::{BTreeMap, BTreeSet};
 use thiserror::Error;
 
+/// The state that is directly recorded in the header.
+///
+/// This state affects the consensus, unlike the ordinary state (which is used for data recording).
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Clone)]
+pub struct EssentialState {
+    /// The original validator set information before the delegation calculation.
+    ///
+    /// The order here is the leader selection priority order of the validators
+    /// which directly affects the result of `calculate_net_validator_set()`.
+    pub validator_set: Vec<(PublicKey, u64)>,
+    /// The protocol version that must be used from next block.
+    ///
+    /// It must be a valid semantic version (e.g., `0.2.3`).
+    pub version: String,
+    /// The delegation state (delegator, delegatee).
+    pub delegation: Vec<(PublicKey, PublicKey)>,
+}
+
+impl EssentialState {
+    /// Calculate the actual set of validators & voting power for the next block.
+    ///
+    /// The order here is same as the order of leaders in each round.
+    /// returns `None` if the delegation is not valid.
+    pub fn calculate_net_validator_set(&self) -> Result<Vec<(PublicKey, u64)>, String> {
+        // check delegation by delegatee (which's forbidden)
+        let mut delegatees = BTreeSet::new();
+        for (delegator, delegatee) in &self.delegation {
```

Comment:
> check delegation by delegatee (which's forbidden).

If we forbid chained delegations, we have to get permission from both delegator and delegatee. Otherwise, delegation to a random node will prevent the node from delegating to another, regardless of its will.
[See](https://github.com/postech-dao/simperby/blob/ebf850ff639bc1c502cfe4e6881eae4c7976b875/simperby/src/lib.rs#L36-L51).

Reply:
Let's revisit this issue later. https://github.com/postech-dao/simperby/issues/49

Reply:
Opened #49